### PR TITLE
apple mps: training support for SDXL (ControlNet, LoRA, Dreambooth, T2I)

### PR DESCRIPTION
**Pull request by bghira**
# What does this PR do?

Adds MPS bugfixes to the SDXL example training scripts.

I haven't really completed this yet, I'm looking for comments on the approach, or whether it should be done at all.

On an M3 Max, I get reasonable speeds for training.

### CODE REVIEW

**Comment by sayakpaul** [comment]
@pcuenca could you give this a look as well?

**Comment by bghira** [comment]
@sayakpaul i've updated the rest of the SDXL training scripts:

* we override --mixed_precision value before creating Accelerator
* we disable autocast when using fp16, bf16, or MPS is enabled. this is to match the scripts to each other, as some already did this and others did not
* the scripts were sometimes slightly different - some use log_validations as a method and others make the images inline

I wanted to make them all more consistent, but I wanted to limit the changes.

File: examples/controlnet/train_controlnet_sdxl.py [hunk 1]
```diff
@@ -792,6 +796,10 @@ def main(args):
 
     logging_dir = Path(args.output_dir, args.logging_dir)
 
+    if torch.backends.mps.is_available():
+        # due to pytorch#99272, MPS does not yet support bfloat16.
+        args.mixed_precision = "fp16"
```
Comment (sayakpaul) [inline examples/controlnet/train_controlnet_sdxl.py:801]: Shouldn't we raise an error here explicitly? I am more in favor of that.

File: examples/dreambooth/train_dreambooth_lora_sdxl.py [hunk 2]
```diff
@@ -205,11 +204,18 @@ def log_validation(
...
+    ):
+        enable_autocast = False
+    if "playground" in args.pretrained_model_name_or_path:
+        enable_autocast = False
 
-    with inference_ctx:
+    with torch.autocast(
+        str(accelerator.device).replace(":0", ""),
+        enabled=enable_autocast,
+    ):
```
Comment (sayakpaul) [inline examples/dreambooth/train_dreambooth_lora_sdxl.py:218]: (nit): Could potentially be made into a single line.

File: examples/dreambooth/train_dreambooth_lora_sdxl.py [hunk 3]
```diff
@@ -205,11 +204,18 @@ def log_validation(
     generator = torch.Generator(device=accelerator.device).manual_seed(args.seed) if args.seed else None
     # Currently the context determination is a bit hand-wavy. We can improve it in the future if there's a better
     # way to condition it. Reference: https://github.com/huggingface/diffusers/pull/7126#issuecomment-1968523051
-    inference_ctx = (
-        contextlib.nullcontext() if "playground" in args.pretrained_model_name_or_path else torch.cuda.amp.autocast()
-    )
+    enable_autocast = True
+    if torch.backends.mps.is_available() or (
+        accelerator.mixed_precision == "fp16" or accelerator.mixed_precision == "bf16"
```
Comment (sayakpaul) [inline examples/dreambooth/train_dreambooth_lora_sdxl.py:209]: Does `torch.cuda.amp.autocast()` translate to 

```python
if accelerator.mixed_precision == "fp16" or accelerator.mixed_precision == "bf16":
    with torch.autocast("cuda", enabled=False):
        ...
```
?

File: examples/dreambooth/train_dreambooth_lora_sdxl.py [hunk 4]
```diff
@@ -959,6 +966,10 @@ def main(args):
     if args.do_edm_style_training and args.snr_gamma is not None:
         raise ValueError("Min-SNR formulation is not supported when conducting EDM-style training.")
 
+    if torch.backends.mps.is_available():
+        # due to pytorch#99272, MPS does not yet support bfloat16.
+        args.mixed_precision = "fp16"
```
Comment (sayakpaul) [inline examples/dreambooth/train_dreambooth_lora_sdxl.py:971]: Same as above.

File: examples/dreambooth/train_dreambooth_lora_sdxl.py [hunk 5]
```diff
@@ -1036,7 +1048,7 @@ def main(args):
                     image.save(image_filename)
 
             del pipeline
-            if torch.cuda.is_available():
+            if accelerator.device.type == "cuda":
```
Comment (sayakpaul) [inline examples/dreambooth/train_dreambooth_lora_sdxl.py:1051]: In one of the earlier instances we did `if torch.cuda.is_available():`. Could we maybe stick to that?

File: examples/dreambooth/train_dreambooth_lora_sdxl.py [hunk 6]
```diff
@@ -1126,6 +1138,10 @@ def main(args):
     elif accelerator.mixed_precision == "bf16":
         weight_dtype = torch.bfloat16
 
+    if torch.backends.mps.is_available():
+        # due to pytorch#99272, MPS does not yet support bfloat16.
+        weight_dtype = torch.float16
```
Comment (sayakpaul) [inline examples/dreambooth/train_dreambooth_lora_sdxl.py:1143]: Same as above.

File: examples/instruct_pix2pix/train_instruct_pix2pix_sdxl.py [hunk 7]
```diff
@@ -981,6 +981,13 @@ def collate_fn(examples):
     if accelerator.is_main_process:
         accelerator.init_trackers("instruct-pix2pix-xl", config=vars(args))
 
+    # Some configurations require autocast to be disabled.
+    enable_autocast = True
+    if torch.backends.mps.is_available() or (
+        accelerator.mixed_precision == "fp16" or accelerator.mixed_precision == "bf16"
+    ):
+        enable_autocast = False
```
Comment (sayakpaul) [inline examples/instruct_pix2pix/train_instruct_pix2pix_sdxl.py:989]: Same comment as above.

**Review by sayakpaul — COMMENTED** [review]
Left some comments, thanks for this.

File: examples/dreambooth/train_dreambooth_lora_sdxl.py (hunk 2 above)
Comment (bghira) [reply examples/dreambooth/train_dreambooth_lora_sdxl.py:218]: the `make style` changes that.

File: examples/dreambooth/train_dreambooth_lora_sdxl.py (hunk 3 above)
Comment (bghira) [reply examples/dreambooth/train_dreambooth_lora_sdxl.py:209]: i'm not sure - but relying on torch.cuda seems counterproductive when it might not be built

File: examples/dreambooth/train_dreambooth_lora_sdxl.py [hunk 8]
```diff
@@ -1001,7 +1012,8 @@ def main(args):
         cur_class_images = len(list(class_images_dir.iterdir()))
 
         if cur_class_images < args.num_class_images:
-            torch_dtype = torch.float16 if accelerator.device.type == "cuda" else torch.float32
+            has_supported_fp16_accelerator = accelerator.device.type == "cuda" or torch.backends.mps.is_available()
+            torch_dtype = torch.float16 if has_supported_fp16_accelerator else torch.float32
```
[the remark this answers is not in the document]
Comment (bghira) [reply examples/dreambooth/train_dreambooth_lora_sdxl.py:1016]: i was updating the code to work for both cuda and mps, and it already does that for cuda

[the remark this answers is not in the document]
Comment (bghira) [reply examples/dreambooth/train_dreambooth_lora_sdxl.py:1016]: you would like me to update it for cuda *and* mps to ask the user?

[the remark this answers is not in the document]
Comment (sayakpaul) [reply examples/dreambooth/train_dreambooth_lora_sdxl.py]: For CUDA and `torch.float16` there's no silent change of the `torch_dtype`. But for MPS, it's the case right? If so, yes, let's explicitly ask the user.

File: examples/dreambooth/train_dreambooth_lora_sdxl.py (hunk 3 above)
Comment (sayakpaul) [reply examples/dreambooth/train_dreambooth_lora_sdxl.py:209]: I see. Could you make it a little more granular then?

```python
if torch.backends.mps.is_available() or "playground" in args.pretrained_model_name_or_path:
    context = torch.autocast(
        str(accelerator.device).replace(":0", ""),
        enabled=False,
    )
elif torch.cuda.is_available():
    context = torch.cuda.amp.autocast()

...
```

WDYT?

File: examples/dreambooth/train_dreambooth_lora_sdxl.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (bghira) [reply examples/dreambooth/train_dreambooth_lora_sdxl.py]: no - this is just for generating more prior loss images.

the original code uses a fp16 dtype for any cuda GPU, and fp32 for CPU

[the remark this answers is not in the document]
Comment (bghira) [reply examples/dreambooth/train_dreambooth_lora_sdxl.py]: it could be updated to use the same dtype as the principle weights but i'd rather keep it the same for this particular set of changes, it's a bit of a pain to change this many files :P

**Comment by bghira** [comment]
@sayakpaul okay,

* we'll toss an error when the user requests bf16 on mps
* i've left it as torch.autocast since relying on the cuda-specific codepath feels incorrect while broadening the support

File: examples/dreambooth/train_dreambooth_lora_sdxl.py [hunk 9]
```diff
@@ -205,11 +204,18 @@ def log_validation(
...
+    if torch.backends.mps.is_available() or (
+        accelerator.mixed_precision == "fp16" or accelerator.mixed_precision == "bf16"
+    ):
+        enable_autocast = False
+    if "playground" in args.pretrained_model_name_or_path:
+        enable_autocast = False
 
-    with inference_ctx:
+    with torch.autocast(
+        str(accelerator.device).replace(":0", ""),
```
Comment (pcuenca) [inline examples/dreambooth/train_dreambooth_lora_sdxl.py:216]: Isn't this `accelerator.device.type`?

File: examples/dreambooth/train_dreambooth_lora_sdxl.py (hunk 3 above)
Comment (bghira) [reply examples/dreambooth/train_dreambooth_lora_sdxl.py:209]: i reviewed the docs and torch.autocast since 1.10 is the canonical method for handling autocast. before this, the standard method was `torch.cuda.amp.autocast()`

but now `torch.autocast("cuda", ...)` and `torch.autocast("cpu", ...)` are the way to go, as they route to their respective `torch.amp.<backend>.autocast()` methods, which means this code also works with the `accelerate config` option for CPU training being selected

File: examples/dreambooth/train_dreambooth_lora_sdxl.py [hunk 10]
```diff
@@ -205,11 +204,18 @@ def log_validation(
...
+        enabled=enable_autocast,
```
Comment (pcuenca) [inline examples/dreambooth/train_dreambooth_lora_sdxl.py:217]: Just curious: why do we disable `autocast` in this script but `contextlib.nullcontext` in others?

File: examples/instruct_pix2pix/train_instruct_pix2pix_sdxl.py [hunk 11]
```diff
@@ -96,7 +91,7 @@ def log_validation(
         else Image.open(image_url_or_path).convert("RGB")
     )(args.val_image_url_or_path)
 
-    with torch.autocast(str(accelerator.device).replace(":0", ""), enabled=accelerator.mixed_precision == "fp16"):
+    with torch.autocast(str(accelerator.device).replace(":0", ""), enabled=enable_autocast):
```
Comment (pcuenca) [inline examples/instruct_pix2pix/train_instruct_pix2pix_sdxl.py:94]: Same comment about `accelerator.device.type`

File: examples/dreambooth/train_dreambooth_lora_sdxl.py (hunk 3 above)
Comment (sayakpaul) [reply examples/dreambooth/train_dreambooth_lora_sdxl.py:209]: Okay beautiful. Thank you.

**Comment by sayakpaul** [comment]
The changes are nice and clean. I am okay. @pcuenca to give this a look!

File: examples/instruct_pix2pix/train_instruct_pix2pix_sdxl.py [hunk 12]
```diff
@@ -981,6 +983,13 @@ def collate_fn(examples):
     if accelerator.is_main_process:
         accelerator.init_trackers("instruct-pix2pix-xl", config=vars(args))
 
+    # Some configurations require autocast to be disabled.
+    enable_autocast = True
+    if torch.backends.mps.is_available() or (
+        accelerator.mixed_precision == "fp16" or accelerator.mixed_precision == "bf16"
```
Comment (pcuenca) [inline examples/instruct_pix2pix/train_instruct_pix2pix_sdxl.py:989]: Is this the same as before? We used autocast most times, didn't we?

**Review by pcuenca — COMMENTED** [review]
Looking good so far!

File: examples/dreambooth/train_dreambooth_lora_sdxl.py (hunk 10 above)
Comment (sayakpaul) [reply examples/dreambooth/train_dreambooth_lora_sdxl.py:217]: This is on me.

File: examples/instruct_pix2pix/train_instruct_pix2pix_sdxl.py (hunk 12 above)
Comment (bghira) [reply examples/instruct_pix2pix/train_instruct_pix2pix_sdxl.py:989]: yep, the logic came from this script

File: examples/dreambooth/train_dreambooth_lora_sdxl.py (hunk 10 above)
Comment (bghira) [reply examples/dreambooth/train_dreambooth_lora_sdxl.py:217]: if you want to peel the can of worms back further, we also need to use

```py
            if is_transformers_available() and transformers.deepspeed.is_deepspeed_zero3_enabled():
                context_manager = deepspeed.zero.GatheredParameters(param, modifier_rank=None)
```

for deepspeed support.

it's starting to feel like we need an autocast manager wrapper in train utils

**Comment by bghira** [comment]
thanks for the help and feedback, this feels ready now.

**Comment by sayakpaul** [comment]
I am good to proceed with the merge. @pcuenca could you also give this a final check?

File: examples/instruct_pix2pix/train_instruct_pix2pix_sdxl.py (hunk 12 above)
Comment (pcuenca) [reply examples/instruct_pix2pix/train_instruct_pix2pix_sdxl.py:989]: Maybe I'm confused, but: https://github.com/huggingface/diffusers/blob/3bce0f3da1c0c13c5589cd97946ddbf58b8a9031/examples/instruct_pix2pix/train_instruct_pix2pix_sdxl.py#L99

**Comment by bghira** [comment]
this is very exciting. i feel like we should make a release note that MPS support in pytorch is preliminary and may result in dramatically different outputs versus a CUDA training run or even using the CPU on the same Apple system.

relevant pytorch bugs:

https://github.com/pytorch/pytorch/issues/109457
https://github.com/pytorch/pytorch/issues/111169
https://github.com/pytorch/pytorch/issues/111634
https://github.com/pytorch/pytorch/issues/117826
https://github.com/pytorch/pytorch/issues/122045
https://github.com/pytorch/pytorch/issues/122123
https://github.com/pytorch/pytorch/issues/122233

**Comment by bghira** [comment]
i tested kohya's sd-scripts on apple mps and it only works with fp32 due to these bugs being unfixed there, so it ends up being a fairly solid implementation to reference if they wish to fix the same.

**Comment by sayakpaul** [comment]
Slightly confused here. Are you saying you are not getting expected results on MPS with these changes and Kohya scripts yield expected results with potentially unfixed issues?

**Comment by bghira** [comment]
no - fp16 and fp32 training do the same thing, but MPS outputs differ from CPU outputs.

in general there are subtle correctness issues in pytorch's mps support that could add up to a different trained result. they aren't worse - just different

**Comment by sayakpaul** [comment]
Sorry it slipped through the cracks. Gonna merge after CI is green.

**Comment by bghira** [comment]
> I reported one of the [Pytorch MPS correctness bugs](https://github.com/pytorch/pytorch/issues/84936). That was well over a year ago, but IIRC the problem is that the MPS implementation of certain transcendental functions, like `exp` and `log`, are not as precise as the CPU implementations.
> 
> From my cursory examination, it appears this patch addresses these bugs by disabling MPS in certain areas. In that sense, this looks like less of a patch to support MPS and more like a patch to disable it. 

sorry, i might have missed something, how is this patch disabling mps? it is ensuring all dtypes are the same as mps requires.

**Comment by mallman** [comment]
> > I reported one of the [Pytorch MPS correctness bugs](https://github.com/pytorch/pytorch/issues/84936). That was well over a year ago, but IIRC the problem is that the MPS implementation of certain transcendental functions, like `exp` and `log`, are not as precise as the CPU implementations.
> > From my cursory examination, it appears this patch addresses these bugs by disabling MPS in certain areas. In that sense, this looks like less of a patch to support MPS and more like a patch to disable it.
> 
> sorry, i might have missed something, how is this patch disabling mps? it is ensuring all dtypes are the same as mps requires.

I was looking at things like the code that throws errors for bf16 when mps is enabled. You're right, this doesn't disable mps, it throws errors where it is unsupported.

At this point, my remaining concern is around the difference in output from cpu versus mps. Should end users know to expect this? And what should they know? Should the message be that "mps output is different but valid" or something else like "mps output is different and invalid"? What are the consequences of the imprecision of the "fast math" functions in MPS used here?

**Comment by bghira** [comment]
i'm not sure. at what point does this issue really become a correctness one? the downstream math becoming meaningless could be simply a human construct. i agree that deterministic results are nice to have, but they result in extreme slowdowns on MPS. all i wanted to do was alert the users that it is possibly an issue. their alternative is to really just not train at all, or pay for third party hardware rental.

**Comment by aradley** [comment]
I'm not sure what you mean by "the downstream math becoming meaningless could be simply a human construct". The issue I've highlighted is not one of deterministic Vs non-determanistic, rather, I have presented a clear issue where MPS is **wrong**. This is not a human construct, the MPS based torch.where function literally gives an incorrect output.

If your patch is not trying to rectify this issue, then fine, but you did mention specifically cite my bug report (#122233).

However, what I would say is that if something as simple as as torch.where on MPS is giving **incorrect** results, I don't see how you can confidently claim that the results on MPS are "different but not wrong" without some comprehensive evidence to back up that claim. On what basis do you claim that the differences observed are down to small errors as opposed to a critical error, as in the torch.where example I present?

**Comment by bghira** [comment]
unfortunately it's not my place or job to sit here and argue semantics about a platform's correctness, the pytorch forum is likely the correct venue from there, and i will not be involved in that discussion. if you would like to address any issues with the pull request or any operational aspects of MPS in the diffusers toolkit, it would be ideal if you would please open a new issue.

**Comment by bghira** [comment]
i've removed the links to pytorch bugs to reduce the occurrence of comments about upstream issues we're not able to resolve. this isn't the place for discussing it, as frustrating as it may be that upstream does not address them.
